Core/Src/main.c

```diff
@@ -87,17 +89,28 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
-
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1); // start the timer
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
+
   while (1)
   {
     /* USER CODE END WHILE */
 
     /* USER CODE BEGIN 3 */
+	  uint16_t adc_val = read_adc(); // get ADC reading
+
+	  float duty_cycle = 0.05 + (0.05 * (adc_val / 1023.0)); // map adc_val to duty cycle in range [0.05, 0.1]
```

Comment:
lets get these magic numbers a home. although I know what they mean, you have to keep in mind those who are looking at this code for the first time. Please use well named const or #define to make this more readable. The calculation itself tho looks good to me.